core/src/main/scala_2.10/shapeless/annotationmacros.scala

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2015-9 Patrick Grandjean
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package shapeless
+
+import scala.reflect.macros.{Context, whitebox}
+
+/*
+ * AbstractAnnotationMacros is declared in folder scala.
+ */
+@macrocompat.bundle
+class AnnotationMacros(override val c: whitebox.Context) extends AbstractAnnotationMacros(c) {
+  import c.universe._
+
+  def collectTypeAnnotation(annTpe: Type, s: Symbol): Option[Tree] = {
+    val construct0 = construct(annTpe)
+
+    s.typeSignature match {
+      case AnnotatedType(annots, _, _) => // the scala 2.10 specific line
+        annots.collectFirst {
+          case ann if ann.tree.tpe =:= annTpe => construct0(ann.tree.children.tail)
+        }
+      case _ => None
+    }
+  }
+}
+
+/*
+ * Workaround for Scala 2.10, which needs an object.
```

Comment:
macrocompat should allow you to keep to the same bundle style as 2.11+ ... can you say what problems you were seeing?

Reply:
If I remove the companion object, I get compilation errors saying "\<macro> is not a member of object shapeless.AnnotationMacros":

```
[error] /Development/shapeless/core/src/main/scala/shapeless/annotation.scala:56:77: value materializeAnnotation is not a member of object shapeless.AnnotationMacros
[error]   implicit def materialize[A, T]: Annotation[A, T] = macro AnnotationMacros.materializeAnnotation[A, T]
[error]                                                                             ^
[error] /Development/shapeless/core/src/main/scala/shapeless/annotation.scala:117:89: value materializeVariableAnnotations is not a member of object shapeless.AnnotationMacros
[error]   implicit def materialize[A, T, Out <: HList]: Aux[A, T, Out] = macro AnnotationMacros.materializeVariableAnnotations[A, T, Out]
[error]                                                                                         ^
[error] /Development/shapeless/core/src/main/scala/shapeless/annotation.scala:171:89: value materializeTypeAnnotations is not a member of object shapeless.AnnotationMacros
[error]   implicit def materialize[A, T, Out <: HList]: Aux[A, T, Out] = macro AnnotationMacros.materializeTypeAnnotations[A, T, Out]
```
That made we believe that the problem was the same for LazyMacrosRef, which is an object in 2.10 and a class in 2.11+.

Reply:
The `Lazy` macros are "special" in all sorts of strange ways.

A closer comparison would be `AnnotationMacros` [here](https://github.com/milessabin/shapeless/blob/master/core/src/main/scala/shapeless/annotation.scala#L120-L213). If you follow that model I think it's fairly likely you can do this without needed any version specific source.

Reply:
I will try to get closer to the original code. But then there is the extraction of `Types.AnnotatedType` which is slightly different between versions:
* 2.10: https://github.com/scala/scala/blob/v2.10.7/src/reflect/scala/reflect/api/Types.scala#L799
* 2.11: https://github.com/scala/scala/blob/v2.11.12/src/reflect/scala/reflect/api/Types.scala#L899

How to go around this?